### PR TITLE
DAOS-7679 vos: Punch max minor epc in vos_obj_array_remove

### PR DESCRIPTION
**Pull request by jolivier23**
Rather than doing an array removal, we can reserve the top minor
epoch for removal.   This allows ec aggregation to call
vos_obj_array_remove with the same epoch.  It will instead
insert a new punch extent using the max minor epoch, thus
overwriting any underlying extents, allowing VOS aggregation to
remove the underlying extents.

Modify the vos iterator to take an option to skip removed extents

Signed-off-by: Jeff Olivier <jeffrey.v.olivier@intel.com>

### CODE REVIEW

File: src/vos/evtree.c [hunk 1]
```diff
@@ -3395,70 +3395,14 @@ int
...
-
-	rect.rc_ex = *ext;
-	rect.rc_epc = epr->epr_hi;
-	rect.rc_minor_epc = EVT_MINOR_EPC_MAX;
+	struct evt_entry_in	entry = {0};
 
-	evt_ent_array_init(&ent_array);
+	entry.ei_rect.rc_ex = *ext;
+	entry.ei_bound = entry.ei_rect.rc_epc = epr->epr_hi;
+	entry.ei_rect.rc_minor_epc = EVT_MINOR_EPC_MAX;
```
Comment (Nasf-Fan) [inline src/vos/evtree.c:3402]: That is good solution, but there is a boundary case that some update via a distributed transaction may have already used the EVT_MINOR_EPC_MAX. So we need to redefine the maximum sub-operations within a distributed transaction to guarantee that EVT_MINOR_EPC_MAX can be reserved for current purpose.

Comment (jolivier23) [reply src/vos/evtree.c:3402]: I added a new definition for this.

**Review by Nasf-Fan — DISMISSED** [review]

**Review by liuxuezhao — APPROVED** [review]
The EC agg test failure probably can be fixed by https://github.com/daos-stack/daos/pull/5864.

**Comment by jolivier23** [comment]
@liuxuezhao I rebased this patch after your patch landed and I ran with daily_regression and ec Features.    Do you see failures that are unexpected?

**Review by wangdi1 — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Pull request merged**